client/src/components/Sidebar/Sidebar.tsx

```diff
@@ -0,0 +1,157 @@
+
+  // Call onExpandCollapse(true) if menu is initialised in collapsed state
+  React.useEffect(() => {
+    if (collapsed && onExpandCollapse) {
+      onExpandCollapse(true);
+    }
+  }, []);
+
+  // 'peeking' is a temporary state to allow the user to peek in the menu while it is collapsed, or hidden.
+  // When peeking is true, the menu renders as if it's not collapsed, but as an overlay instead of occupying
+  // space next to the content
+  const [peeking, setPeeking] = React.useState(false);
+
+  // Whether or not to display the menu with slim layout.
+  // Separate from 'collapsed' as the menu can still be displayed with an expanded
+  // layout while in 'collapsed' mode if the user is 'peeking' into it (see above)
+  const slim = collapsed && !peeking;
+
+  // 'expandingOrCollapsing' is set to true whilst the the menu is transitioning between slim and expanded layouts
+  const [expandingOrCollapsing, setExpandingOrCollapsing] = React.useState(false);
+  React.useEffect(() => {
+    setExpandingOrCollapsing(true);
+    const finishTimeout = setTimeout(() => {
+      setExpandingOrCollapsing(false);
+    }, 300);
+
+    return () => {
+      clearTimeout(finishTimeout);
+    };
+  }, [slim]);
+
+  const onClickCollapseToggle = (e: React.MouseEvent) => {
+    e.preventDefault();
+    setCollapsed(!collapsed);
+
+    // Unpeek if the user has just collapsed the menu
+    // Otherwise the menu would just stay open until the mouse leaves
+    if (!collapsed) {
+      setPeeking(false);
+    }
+
+    if (onExpandCollapse) {
+      onExpandCollapse(!collapsed);
+    }
+  };
+
+  // Set up an event handler to switch peeking on/off when the mouse cursor hovers the sidebar
+  const wrapperRef = React.useRef<HTMLDivElement | null>(null);
+  // eslint-disable-next-line consistent-return
+  React.useEffect(() => {
+    if (wrapperRef.current) {
+      const element = wrapperRef.current;
+      let startPeekingTimeout;
+      let stopPeekingTimeout;
+
+      const onMouseEnterHandler = () => {
+        clearTimeout(startPeekingTimeout);
+        clearTimeout(stopPeekingTimeout);
+        startPeekingTimeout = setTimeout(() => {
+          setPeeking(true);
+        }, 100);
+      };
+
+      const onMouseLeaveHandler = () => {
+        clearTimeout(startPeekingTimeout);
+        clearTimeout(stopPeekingTimeout);
+        stopPeekingTimeout = setTimeout(() => {
+          setPeeking(false);
+        }, 300);
+      };
+
+      element.addEventListener('mouseenter', onMouseEnterHandler);
+      element.addEventListener('mouseleave', onMouseLeaveHandler);
+
+      return () => {
+        clearTimeout(startPeekingTimeout);
+        clearTimeout(stopPeekingTimeout);
+        element.removeEventListener('mouseenter', onMouseEnterHandler);
+        element.removeEventListener('mouseleave', onMouseLeaveHandler);
+      };
+    }
+  }, [wrapperRef.current]);
```

Comment:
Question:  I am not sure of this approach. Does the `wrapperRef` change over time ? Or you just want to run this once the component has been mounted ?
If so, you can remove the ref and simply write the useEffect this way
```tsx
useEffect(() => {
  // This function will run on first component mount
}, [0])

Reply:
I didn't realise `useEffect` only ran after rendering. Assumed it ran just before which is why I added this here!

Wondering what the ``0`` means here? is this different to passing an empty array?

Reply:
It run after rendering, and after each rendering. I think in your case, it would mean that every update of the `ref` will cause the hook to run. 
The `0` was mandatory, and it is an old habit, but it appears it is not anymore. `[]` is fine